zboxcore/sdk/multi_operation_worker.go

```diff
@@ -0,0 +1,176 @@
+package sdk
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/0chain/errors"
+	// "github.com/0chain/gosdk/constants"
+
+	"github.com/0chain/gosdk/core/util"
+	"github.com/0chain/gosdk/zboxcore/allocationchange"
+	"github.com/0chain/gosdk/zboxcore/client"
+	"github.com/0chain/gosdk/zboxcore/fileref"
+	l "github.com/0chain/gosdk/zboxcore/logger"
+
+	"github.com/0chain/gosdk/zboxcore/zboxutil"
+	"github.com/google/uuid"
+)
+
+type Operationer interface {
+	Process(allocObj *Allocation, connectionID string, totalOperation int) ([]fileref.RefEntity, error)
+	buildChange(refs []fileref.RefEntity, uid uuid.UUID) []allocationchange.AllocationChange
+	Completed(allocObj *Allocation)
+	Error(allocObj *Allocation, consensus int, err error)	
+}
+
+type MultiOperation struct {
+	connectionID  string
+	operations    []Operationer
+	allocationObj *Allocation
+	ctx           context.Context
+	ctxCncl       context.CancelFunc
+	operationMask zboxutil.Uint128
+	maskMU        *sync.Mutex
+	Consensus
+
+	changes [][]allocationchange.AllocationChange
+}
+
+func (mo *MultiOperation) Process() error {
+	l.Logger.Info("MultiOperation Process start")
+	wg := &sync.WaitGroup{}
+	mo.changes = make([][]allocationchange.AllocationChange, len(mo.operations))
+	ctx, ctxCncl := context.WithCancel(context.Background())
+	defer ctxCncl()
+	errs := make(chan error, 1)
+	uid := util.GetNewUUID()
+	for idx, op := range mo.operations {
+		// Don't use goroutine for the first operation because in blobber code we try to fetch the allocation
+		// from the postgress and sharders, if not found blobber try to create it. This is done without lock, so if we 
+		// sent multiple goroutine together, blobber will try to create multiple allocations for same allocation id
+		// and eventually throw error.
```

Comment:
Please protect this with lock in blobber and remove condition checking here

Reply:
Why do you think it should be mutual exclusion? I believe we can lock allocation for a particular group of operations and do them in parallel, just execute them atomically

Reply:
> Why do you think it should be mutual exclusion?

Sorry for the confusion but I was saying we don't require mutual exclusion there.  

> I believe we can lock allocation for a particular group of operations

AFAIK, we can't. Blobbers don't accept a group of operations (@lpoli  can confirm). So, if we are sending multiple operations concurrently, and if we lock the allocation, then the first operation that reaches the blobber will first complete the operation and then allow the next operation to work. It will become mutual exclusion (although we don't need it). It will degrade the performance. That's why I don't want to put lock on allocation.